File: mbeps1/ReadFiles/pfvreadf1.py
#-----------------------------------------------------------------------
# This program reads real 1d velocity data
# written for 1D OpenMP PIC codes
# written by Viktor K. Decyk, UCLA
from __future__ import print_function
import sys
import math
import numpy

sys.path.append('./ReadFiles')
from fgraf import *
from libcmfield1 import *

if (sys.version_info.major==2):
   input = raw_input

#-----------------------------------------------------------------------
def main(idrun):
   """
   reads real periodic 1d velocity data
   idrun = run identifier for current run, -1 if unknown
   """
   int_type = numpy.int32
   double_type = numpy.float64
   float_type = numpy.float32
   complex_type = numpy.complex64

   ns = 2
   iudm = 19; iuv = 12
   dname = numpy.array(["ELECTRON VELOCITY   ","ION VELOCITY        "],
                       dtype=str)
   sname = numpy.array(["ELECTRON","ION     "],dtype=str)

# create string from idrun
   if (idrun < 0):
      cdrun = "Unknown"
      while (cdrun.isdigit() == False):
         cdrun = input("enter integer idrun: ")
      idrun = int(cdrun)
   cdrun = str(idrun)
   fname = "diag1." + cdrun
   cmfield1.ffopen1(iudm,fname)

# nscalars = table of available diagnostics
   nscalars = numpy.zeros((ns),int_type,'F')

# determine which velocity diagnostics are available
   cmfield1.readfvdiags1(iudm,nscalars)

   nts = numpy.zeros((1),int_type,'F')
   nmv = numpy.zeros((1),int_type,'F')
   nfvd = numpy.zeros((1),int_type,'F')
   nfed = numpy.zeros((1),int_type,'F')
   mrec = numpy.zeros((1),int_type,'F')
   fname = numpy.array([""],'S32')
   gname = numpy.array([""],'S32')
   ierr = numpy.zeros((1),int_type,'F')

# open graphics device
   nplot = 4
   ierr[0] = graf1.open_graphs(nplot)

# select diagnostic
   m = numpy.sum(nscalars)
   while True:
      if (m > 0):
         n = -1
         while True:
            if (n < 0):
               for i in range(0,ns):
                  if (nscalars[i]==1):
                     print ("enter ", i+1," for", 
                            numpy.str.rstrip(dname[i]))
               print ("enter ", 0," for EXIT")
               c = input("")
               if (c.isdigit()):
                  n = int(c)
               if (n==0):
                  break
               if ((n >= 1) and (n <= ns)):
                  if (nscalars[n-1]==0):
                     n = -1
            else:
               n = -1
            if (n > 0):
               break
            print ("invalid entry, try again or enter 0 to quit")
      else:
         print ("no velocity-space diagnostic files found")
         n = 0
# exit procedure
      if (n==0):
         if ("fvm" in globals()):
            fvm = None
         if ("fv" in globals()):
            fv = None
         if ("fe" in globals()):
            fe = None
         if ("wk" in globals()):
            wk = None
         cmfield1.closeff1(iudm)
         graf1.close_graphs()
         return

      tname = numpy.str.rstrip(dname[n-1])
      print (tname, " diagnostic selected")

# return parameters for selected velocity diagnostic:
# nts, nmv, nfvd, nfed, nrec, fname
      cmfield1.fvdiagparams1(iudm,n,nts,nmv,nfvd,nfed,mrec,fname)
      nrec = mrec[0]

# nmv21 = number of velocity bins
      nmv21 = 2*nmv[0] + 1; nmvf = nmv21 + 1

# allocate velocity distribution arrays
      if ("fvm" not in globals()):
         fvm = numpy.empty((in1.ndim,3),float_type,'F')
      if ("fv" not in globals()):
         fv = numpy.empty((nmvf,nfvd[0]),float_type,'F')
      if ("fe" not in globals()):
         fe = numpy.empty((nmvf,nfed[0]),float_type,'F')
      if ("wk" not in globals()):
         wk = numpy.empty((1),float_type,'F')
      dt = in1.dt*float(nts[0])

# open stream file for distribution field
      cmfield1.fsopen1(iuv,fname)

# nrec = number of complete records
      print ("records found: nrec = ", nrec)

# read and display data
      for ii in range(0,nrec):
# read real velocity fields
         cmfield1.freadfv1(iuv,fvm,fv,fe,wk,in1.ndim,nmvf,nfvd,nfed)
         it = nts[0]*ii
         time = dt*float(ii)
         if (nfvd[0] > 0):
# display cylindrical distribution
            if ((in1.ndim==3) and (nfvd[0] < in1.ndim)):
               gname = numpy.str.rstrip(sname[n-1])
               graf1.displayfvb1(fv,fvm,gname,it,nmv,2,ierr)
               if (ierr[0]==1):
                  break
# display cartesian distribution
            else:
               gname = numpy.str.rstrip(sname[n-1])
               graf1.displayfv1(fv,fvm,gname,it,nmv,2,ierr)
               if (ierr[0]==1):
                  break
# display energy distribution
            if (nfed[0] > 0):
               gname = numpy.str.rstrip(sname[n-1])
               graf1.displayfe1(fe,wk,gname,it,nmv,ierr)
               if (ierr[0]==1):
                  break

      graf1.reset_graphs()
      graf1.reset_nplot(4,ierr)
      cmfield1.closeff1(iuv)
      print()

if (__name__=="__main__"):
   main(-1)

File: mbeps1/ReadFiles/Python2/pfvreadf1.f90
!-----------------------------------------------------------------------
! This program reads real periodic 1d velocity distribution data
! written for 1D OpenMP PIC codes
! written by Viktor K. Decyk, UCLA
      program pfvreadf1
      use graf1
!
      implicit none
      integer, parameter :: nmv = 80, ndim = 3, ius = 64
      integer :: ii, it, is, ntime, nmv21, ierr
      real :: ci = 0.0
      real :: ci2, fmin, de, e, gam
      double precision :: dsum
      real, dimension(:,:), allocatable :: fv, dfv
      real, dimension(:), allocatable :: g
      character(len=6) :: chr
      character(len=10) :: cdrun
      character(len=32) :: fname
      character(len=10), dimension(2) :: chs = (/'INITIAL   ',          &
     &'FINAL     '/)
!     character(len=10), dimension(3) :: chs = (/'INITIAL   ',          &
!    &'FINAL     ','POST FINAL'/)
!
      nmv21 = 2*nmv + 1
      allocate(fv(nmv21+1,ndim),dfv(nmv21,2),g(nmv21))
!     allocate(fv(nmv21+1,ndim),dfv(nmv21,3),g(nmv21))
      ci2 = ci*ci
!
      write (cdrun,'(i10)') ius
      fname = 'fort.'//adjustl(cdrun)
      open(unit=ius,file=fname,form='formatted',status='old')
!
      ierr = open_graphs(1)

!     fname = 'LOG(GAMMA-1): WATERBAG/JUTTNER'
!     fname = 'LOG(GAMMA-1): MAXWELLIAN/JUTTNER'
      fname = 'LOG(GAMMA-1): JUTTNER/JUTTNER'
      do ii = 1, 1000
!     do ii = 1, 4000
!     do ii = 1, 1500
      read (ius,*) chr, ntime
      dsum = 0.0d0
      do it = 1, size(fv,1)
      read (ius,*) is,fv(it,1)
      if (it.gt.1) dsum = dsum + dble(fv(it,1))
      enddo
      if (ii.eq.1) write (*,*) 'sum=',real(dsum)
!
      if (ii==1) then
         de = fv(1,1)/real(nmv)
         do it = 1, nmv21
         e = de*(real(it)-0.5)
         gam = 1.0 + e*ci2
!        g(it) = gam/sqrt((gam+1.0)*e)
         g(it) = gam*sqrt((gam+1.0)*e)
         enddo
      endif
!     fmin = minval(fv(2:,1),fv(2:,1).gt.0.0)
      fmin = 1
      where (fv(2:,1) > 0.0)
         fv(2:,1) = alog(fv(2:,1)/g)
      else where
         fv(2:,1) = alog(fmin)
      end where
      if (ii==1) dfv(:,1) = fv(2:,1)
      if (ii==1000) dfv(:,2) = fv(2:,1)
!     if (ii==4000) dfv(:,2) = fv(2:,1)
!     if (ii==1500) dfv(:,3) = fv(3:,1)
      call dscaler1(fv(2:,1),fname,ntime,999,1,nmv21,ierr)
      if (ierr==1) exit
      enddo
!
      call dmscaler1(dfv,fname,ntime,999,1,nmv21,chs,ierr)
!
      write (50,*) 'Initial'
      do it = 1, nmv21
      write (50,*) it,dfv(it,1),exp(dfv(it,1)),0.5*g(it)**2
      enddo
!
      write (50,*) 'Final'
      do it = 1, nmv21
      write (50,*) it,dfv(it,2),exp(dfv(it,2)),0.5*g(it)**2
      enddo
!
!     write (50,*) 'Post Final'
!     do it = 1, nmv21
!     write (50,*) it,dfv(it,3),exp(dfv(it,3)),0.5*g(it)**2
!     enddo
!
      read (5,*) chr
!
      call close_graphs()
!
      end program
